src/inference/ess.jl

```diff
@@ -0,0 +1,132 @@
+"""
+    ESS
+
+Elliptical slice sampling algorithm.
+
+# Examples
+```jldoctest; setup = :(Random.seed!(1))
+julia> @model gdemo(x) = begin
+           m ~ Normal()
+           x ~ Normal(m, 0.5)
+       end
+gdemo (generic function with 2 methods)
+
+julia> sample(gdemo(1.0), ESS(), 1_000) |> mean
+Mean
+
+│ Row │ parameters │ mean     │
+│     │ Symbol     │ Float64  │
+├─────┼────────────┼──────────┤
+│ 1   │ m          │ 0.811555 │
+```
+"""
+struct ESS{space} <: InferenceAlgorithm end
+
+ESS() = ESS{()}()
+ESS(space::Symbol) = ESS{(space,)}()
+
+mutable struct ESSState{V<:VarInfo} <: AbstractSamplerState
+    vi::V
+end
+
+ESSState(model::Model) = ESSState(VarInfo(model))
+
+function Sampler(alg::ESS, model::Model, s::Selector)
+    # sanity check
+    space = getspace(alg)
+    if isempty(space)
+        pvars = get_pvars(model)
+        length(pvars) == 1 ||
+            error("[ESS] no symbol specified to sampler although there is not exactly one model parameter ($pvars)")
+    end
+
+    state = ESSState(model)
+    info = Dict{Symbol, Any}()
+
+    return Sampler(alg, info, s, state)
+end
+
+# always accept in the first step
+function step!(::AbstractRNG, ::Model, spl::Sampler{<:ESS}, ::Integer; kwargs...)
+    return Transition(spl)
+end
+
+function step!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:ESS},
+    ::Integer,
+    ::Transition;
+    kwargs...
+)
+    # recompute log-likelihood in logp
+    vi = spl.state.vi
+    if spl.selector.tag !== :default
+        runmodel!(model, vi, spl)
+    end
+
+    # obtain previous sample
+    f = copy(vi[spl])
+
+    # sample log-likelihood threshold for the next sample
+    threshold = getlogp(vi) - randexp(rng)
+
+    # sample from the prior
+    runmodel!(model, vi, spl)
+    ν = copy(vi[spl])
```

Comment:
Again, is `runmodel!` correct here? I just want to obtain a new sample from the prior.

Reply:
In this case you can just use `runmodel!(model, vi)`

Reply:
Again, I need `spl` since I want the log likelihood?

Reply:
If you just want the likelihood, you might want to wait for the compiler3.0 PR which makes the likelihood computation easier and more efficient.